pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerMemoryLeakTest.java

```diff
@@ -0,0 +1,370 @@
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING)
+                .blockIfQueueFull(false).maxPendingMessages(1)
+                .enableBatching(true).topic(topicName).create();
+        List<MsgPayloadTouchableMessageBuilder<String>> msgBuilderList = new ArrayList<>();
+        for (int i = 0; i < 100; i++) {
+            msgBuilderList.add(newMessage(producer));
+        }
+        List<CompletableFuture> sendFutureList = new ArrayList<>();
+        for (MsgPayloadTouchableMessageBuilder<String> msgBuilder: msgBuilderList) {
+            sendFutureList.add(msgBuilder.value("msg-1").sendAsync());
+        }
+        try{
+            sendFutureList.get(sendFutureList.size() - 1).join();
+        } catch (Exception ex) {
+            // Ignore the error PulsarClientException$ProducerQueueIsFullError.
+            assertTrue(FutureUtil.unwrapCompletionException(ex)
+                    instanceof PulsarClientException.ProducerQueueIsFullError);
+        }
+
+        // Verify: ref is expected.
+        producer.close();
+        for (int i = sendFutureList.size() - 1; i > -1; i--) {
+            MsgPayloadTouchableMessageBuilder<String> msgBuilder = msgBuilderList.get(i);
+            assertEquals(msgBuilder.payload.refCnt(), 1);
+            msgBuilder.release();
+            assertEquals(msgBuilder.payload.refCnt(), 0);
+        }
+        for (int i = sendFutureList.size(); i < 100; i++) {
+            MsgPayloadTouchableMessageBuilder<String> msgBuilder = msgBuilderList.get(i);
+            msgBuilder.release();
+            assertEquals(msgBuilder.payload.refCnt(), 0);
+        }
+        admin.topics().delete(topicName);
+    }
+
+    /**
+     * The content size of msg(value is "msg-1") will be "5".
+     * Then provides two param: 1 and 5.
+     *   1: reach the limitation before adding the message metadata.
+     *   2: reach the limitation after adding the message metadata.
+     */
+    @DataProvider(name = "maxMessageSizeAndCompressions")
+    public Object[][] maxMessageSizeAndCompressions(){
+        return new Object[][] {
+                {1, CompressionType.NONE},
+                {5, CompressionType.NONE},
+                {1, CompressionType.LZ4},
+                {6, CompressionType.LZ4}
+        };
+    }
+
+    @Test(dataProvider = "maxMessageSizeAndCompressions")
+    public void testSendMessageSizeExceeded(int maxMessageSize, CompressionType compressionType) throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        admin.topics().createNonPartitionedTopic(topicName);
+        ProducerImpl<String> producer = (ProducerImpl<String>) pulsarClient.newProducer(Schema.STRING).topic(topicName)
+                .compressionType(compressionType)
+                .enableBatching(false)
+                .create();
+        final ClientCnx cnx = producer.getClientCnx();
```

Comment:
I don't get why we need to get `cnx` here. It seems it only calls `cnx.close()` in the following cide.